azure/azure.go

```diff
@@ -46,11 +58,112 @@ var (
 
 type azureOps struct {
 	cloudops.Compute
-	instance          string
-	resourceGroupName string
-	disksClient       *compute.DisksClient
-	vmsClient         vmsClient
-	snapshotsClient   *compute.SnapshotsClient
+	instance           string
+	resourceGroupName  string
+	managedClusterName string
+	agentPoolName      string
+	disksClient        *compute.DisksClient
+	vmsClient          vmsClient
+	snapshotsClient    *compute.SnapshotsClient
+	agentPoolsClient   *containerservice.AgentPoolsClient
+}
+
+// NewAzureClient initializes cloudops driver for azure based on environment
+// variables or based on instance metadata info available inside Azure VM
+func NewAzureClient() (cloudops.Ops, error) {
+
+	if onAzure() {
+		logrus.Info("Running on Azure IaaS VM")
+
+		client := &http.Client{}
+
+		req, _ := http.NewRequest("GET", metadataAPIEndpoint, nil)
+		req.Header.Add("Metadata", "True")
+		q := req.URL.Query()
+		q.Add("format", "json")
+		q.Add("api-version", metadataAPIVersion)
+		req.URL.RawQuery = q.Encode()
+
+		resp, err := client.Do(req)
+		if err != nil {
+			logrus.Errorf("Error occured while getting instance metadata from Azure Metadata API. Error:[%v]", err)
+			return nil, err
+		}
+
+		defer resp.Body.Close()
+		if resp.StatusCode != 200 {
+			return nil, fmt.Errorf("Error querying Azure metadata: Code %d returned for url %s", resp.StatusCode, req.URL)
+		}
+		respBody, _ := ioutil.ReadAll(resp.Body)
+
+		if len(respBody) == 0 {
+			return nil, fmt.Errorf("Error querying Azure metadata: Empty response")
+		}
+
+		metadata := make(map[string]interface{})
+		err = json.Unmarshal(respBody, &metadata)
+		if err != nil {
+			return nil, fmt.Errorf("Error parsing Azure metadata: %v", err)
+		}
+
+		var vmID, scalesetName, subscriptionID, resourceGroupName, managedClusterName, agentPoolName string
+		var scalesetNametemp, vmIDtemp, subscriptionIDtemp interface{}
+		var exists bool
+
+		if resourceGroup, ok := metadata[resourceGroupNameKey]; ok {
+			temp := resourceGroup.(string)
+			if len(strings.Split(temp, "_")) == 2 {
+				resourceGroupName = strings.Split(temp, "_")[1]
+				managedClusterName = strings.Split(temp, "_")[2]
```

Comment:
unless i'm missing something, this will panic. `if len(strings.Split(temp, "_")) == 2` means where are 2 items in the list so you cannot index `strings.Split(temp, "_")[2]`.